geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/reader/StreamingSimpleFeatureArrowFileReader.scala

```diff
@@ -0,0 +1,143 @@
+* Copyright (c) 2013-2017 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.arrow.io.reader
+
+import java.io.{Closeable, InputStream}
+
+import org.apache.arrow.memory.BufferAllocator
+import org.apache.arrow.vector.complex.NullableMapVector
+import org.apache.arrow.vector.stream.ArrowStreamReader
+import org.locationtech.geomesa.arrow.features.ArrowSimpleFeature
+import org.locationtech.geomesa.arrow.io.SimpleFeatureArrowFileReader.{SkipIndicator, VectorToIterator}
+import org.locationtech.geomesa.arrow.io.{SimpleFeatureArrowFileReader, SimpleFeatureArrowIO}
+import org.locationtech.geomesa.arrow.vector.{ArrowDictionary, SimpleFeatureVector}
+import org.opengis.feature.simple.SimpleFeatureType
+import org.opengis.filter.Filter
+
+import scala.collection.mutable.ArrayBuffer
+
+/**
+  * Streams features from an input stream - note that a feature may not be valid after a call to `.next()`,
+  * as the underlying data may be reclaimed
+  *
+  * @param is creates a new input stream
+  * @param allocator buffer allocator
+  */
+class StreamingSimpleFeatureArrowFileReader(is: () => InputStream)(implicit allocator: BufferAllocator)
+    extends SimpleFeatureArrowFileReader  {
+
+  private var initialized = false
+
+  private lazy val metadata = {
+    initialized = true
+    new StreamingSingleFileReader(is())
+  }
+
+  // TODO read sort without loading a batch...
+  private lazy val sort = SimpleFeatureArrowIO.getSortFromMetadata(metadata.metadata)
+
+  override def sft: SimpleFeatureType = metadata.sft
+
+  override def dictionaries: Map[String, ArrowDictionary] = metadata.dictionaries
+
+  override def features(filter: Filter): Iterator[ArrowSimpleFeature] with Closeable = {
+    val stream = is()
+    val skip = new SkipIndicator
+    val nextBatch = SimpleFeatureArrowFileReader.features(sft, filter, skip, sort, dictionaries)
+
+    // we track all the readers and close at the end to avoid closing the input stream prematurely
+    val readers = ArrayBuffer.empty[StreamingSingleFileReader]
+    // reader for current logical 'file'
+    var reader: StreamingSingleFileReader = null
+
+    new Iterator[ArrowSimpleFeature] with Closeable {
+      private var done = false
+      private var batch: Iterator[ArrowSimpleFeature] = Iterator.empty
+
+      override def hasNext: Boolean = {
+        if (done) {
+          false
+        } else if (batch.hasNext) {
+          true
+        } else if (!skip.skip && stream.available() > 0) {
+          // new logical file
+          reader = new StreamingSingleFileReader(stream)
+          readers.append(reader)
+          batch = reader.features(nextBatch, skip)
```

Comment:
yeah, you could do that - you mean so that we don't have to load the metadata separately? one problem with that is you have to keep the reader open and not close it, or the dictionaries may get invalidated. there are some improvements we could make here (which is why i put the TODO), but they're not entirely trivial. a lot of the underlying clumsiness comes from the fact that the arrow reader can't just read the metadata/dictionaries and stop, it has to load the first batch at the same time. i'd like to fix that in arrow first, then this logic should get easier